kernel/src/net/udp_port_table.rs

```diff
@@ -0,0 +1,293 @@
+//! In-kernel structure for tracking UDP ports bound by capsules.
```

Comment:
Foreword: if this table is going to be in the kernel, then I think it is worth making sure we are happy with the interface and implementation, and that could potentially take some discussion.

With that disclaimer, my first question is why is this in the kernel (I use "kernel" to mean "kernel crate")? I ask for two reasons: 1) the code that enforces that the port bindings are valid is not in the kernel, and 2) ports provided to userspace are not checked by this code. Given both of those, why not just leave port management up to the capsules in `capsules/net`? Port management certainly could be a kernel level operation, but for that to be really effective it then also has to enforce the bindings (i.e. route incoming packets based on the bindings).

If this should be in the kernel, then why does it not manage userland ports as well? Perhaps something could request a range of ports for userland. I do sympathize with the complexity of writing code where both capsules and userland processes share resources, however. I don't think it is very natural at the moment, and I don't think we have a standard way to do it.

If this should be in the kernel, then I think changing some of the names would be worthwhile.
-`PORT_TABLE` => `USED_KERNEL_PORTS`
-`UdpPortSocket` => `UdpSocket`
-`UdpPortTable` => `UdpPortManager`
-`PortEntry` => `SocketBindingEntry`
-`UdpReceiverBinding` => `UdpPortBindingReceive`

Also hopefully others will add comments as well...

Reply:
Thanks for taking time to go through all of this!

Thinking about it now, I think it would be possible to move this into `capsules/net/udp`. Originally this was all in the kernel because we were keeping the door open  toward moving some portion of sending/receiving UDP packets into the kernel such that bindings could be managed and enforced in the kernel, and other potential network security considerations could be enforced in the kernel. 

In hindsight, it seems unlikely that we will move enforcement into the kernel anytime soon. Given this, I think having `UdpPortTable::new()` require some capability to be called would have basically the same effect as our current implementation if `udp_port_table.rs` was in `capsules/net/udp`. This is a reasonable approach for now, I'll make that change.

Re: why does this not manage userland ports as well?
A: There are several advantages to storing userland ports in the grant region of each app. First, grant regions are dynamically allocated / removed when apps are added / removed, so storing userland ports in the grant region scales to any number of apps without the kernel having to guess as to how many userland apps will bind to ports. Second, this approach ensures that whenever an app is removed from a board any ports it has bound are immediately unbound, as the removal of the grant region effectively removes the binding -- enforcing this otherwise is difficult, as `drop()` is not called on objects in the grant region because we simply zero out that memory when apps are removed. This second point is important, as we don't want a design where an app failing locks a replacement of that app out from a particular port it wanted to use. Third, the current design relies on each capsule holding a single `UdpSender` / `UdpReceiver` for each port it binds to, and dynamically allocating those objects for each new app sounds difficult. The two-tier virtualization in this PR also makes it easy to prioritize capsules over userspace apps when it comes to sending queued packets.

Even though I am going to move the port table to capsules, I'll still apply the name changes, they are definitely an improvement.